### PR TITLE
Ignoring empty ssh keys when updating groups

### PR DESCRIPTION
**Pull request by ffilippopoulos**

### CODE REVIEW

File: sync.go [hunk 1]
```diff
@@ -55,10 +55,33 @@ type Group struct {
...
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(body)
+	b := buf.Bytes()
+	json.Unmarshal(b, &adminUser)
+
+	if len(adminUser.CustomSchemas.Keys.SSH) > 0 {
+		group.Keys = append(group.Keys, adminUser.CustomSchemas.Keys.SSH)
+	} else {
+		log.Printf("Ignoring empty key for member: %s group: %s", m.Email, group.Name)
```
Comment (george-angel) [inline sync.go]: Thinking about it, I don't see this log message being valuable, its fairly obvious behaviour of what happens if your key isn't set. So we can probably lose then `else` bit, what do you think?

Comment (ffilippopoulos) [reply sync.go]: Makes sense, that would actually remove the useless `member` argument from the function as well. It only serves the case where you don't see your key being populated under /authmap and you want to trace it via logs, which is pointless for the case of an empty key. Removing..

Comment (george-angel) [reply sync.go]: Ta ^^

File: ssh_key_manager_test.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strings"
+	"testing"
+)
+
+func encodeToReader(i interface{}) io.ReadCloser {
```
Comment (catalin-ilea) [inline ssh_key_manager_test.go]: Can probably simplify this, something like

```
var b bytes.Buffer
 json.NewEncoder(&b).Encode(i)
 return ioutil.NopCloser(&b)
```

File: ssh_key_manager_test.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	b := new(bytes.Buffer)
+	enc := json.NewEncoder(b)
+	enc.Encode(i)
+
+	return ioutil.NopCloser(strings.NewReader(b.String()))
+}
+
+func TestDecodeMemberList(t *testing.T) {
+	gMember1 := &GoogleMember{"member1@uw.co.uk"}
```
Comment (catalin-ilea) [inline ssh_key_manager_test.go]: Why using pointers here(and below)? It just makes the tests more verbose.

File: ssh_key_manager_test.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	gMember2 := &GoogleMember{"member2@uw.co.uk"}
+	gMemberList := &GoogleMemberList{
+		Members: []GoogleMember{*gMember1, *gMember2},
+	}
+	r := encodeToReader(*gMemberList)
+
+	memberList, err := decodeMemberList(r)
+	if err != nil {
+		log.Fatal(err)
```
Comment (catalin-ilea) [inline ssh_key_manager_test.go]: You should probably use t.Fatal instead

File: ssh_key_manager_test.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	}
+
+	if len(memberList.Members) != 2 {
+		t.Error("Unexpected MemberList", memberList.Members)
```
Comment (catalin-ilea) [inline ssh_key_manager_test.go]: Again, you should use t.Fatal instead so that you stop execution at this point and fail fast

File: sync.go [hunk 6]
```diff
@@ -55,10 +55,31 @@ type Group struct {
...
+	dec := json.NewDecoder(body)
+	err := dec.Decode(&memList)
+
+	return memList, err
+}
+
+func (group *Group) addSSHKeys(body io.ReadCloser) {
+	var adminUser GoogleAdminUser
+
+	buf := new(bytes.Buffer)
```
Comment (catalin-ilea) [inline sync.go]: You can simplify this by using a Decoder as you do in `decodeMemberList()`

File: sync.go [hunk 7]
```diff
@@ -55,10 +55,31 @@ type Group struct {
 	Keys []string `json:"keys"`
 }
 
+func decodeMemberList(body io.ReadCloser) (GoogleMemberList, error) {
```
Comment (catalin-ilea) [inline sync.go]: Why ReadCloser instead of io.Reader? You're not closing the body in these functions

**Comment by catalin-ilea** [comment]
Looks good overall

**Comment by ffilippopoulos** [comment]
Made some changes to simplify stuff after Catalin's comments

**Pull request merged**